### ISSUE TITLE
[Feature Request]: refactor _update_causal_mask to  a public utility

### ISSUE THREAD

**Issue by Irvingwangjr** [huggingface/transformers#36640]
### Feature request

 refactor _update_causal_mask to  a public utility

### Motivation

After this pr https://github.com/huggingface/transformers/pull/35235/files#diff-06392bad3b9e97be9ade60d4ac46f73b6809388f4d507c2ba1384ab872711c51
all the attention implement already refactor to use ALL_ATTENTION_FUNCTIONS and people can register their own implement very easy.

I notice that there are still another function: _update_causal_mask  is copy-and-paste everywhere and related to attention modules

if people register an attention impl, this _update_causal_mask will add attention_mask if it is not flash_attention_2. so hope this function can be refactor too

### Your contribution

I can do some testing and submitting a PR, we can add ulysess implementation as a third party example

**Comment by Rocketknight1** [comment]
cc @cyrilvallez

**Comment by Cyrilvallez** [comment]
Hey @Irvingwangjr! It is actually on the roadmap, and I will start working on it very soon! 🤗 Not yet opened a PR, but mask creation will be refactored for sure 😉
